File: backend/src/routes/moderation.js
import { Router } from 'express';
import { query } from '../db/pool.js';
import { authenticate } from '../middleware/auth.js';

const router = Router();

async function requireAdmin(req, res, next) {
    try {
        const result = await query('SELECT is_admin FROM users WHERE id = $1', [req.user.id]);
        if (!result.rows[0]?.is_admin) {
            return res.status(403).json({ error: 'Admin access required' });
        }
        next();
    } catch (err) {
        next(err);
    }
}

router.get('/queue', authenticate, requireAdmin, async (_req, res, next) => {
    try {
        const [posts, echoes, reports] = await Promise.all([
            query(`SELECT id, title, type, status, created_at FROM posts WHERE status = 'pending' ORDER BY created_at ASC`),
            query(`SELECT id, post_id, body, status, created_at FROM echoes WHERE status = 'pending' ORDER BY created_at ASC`),
            query(`SELECT id, target_type, target_id, reason, status, created_at FROM reports WHERE status = 'open' ORDER BY created_at ASC`),
        ]);
        res.json({ posts: posts.rows, echoes: echoes.rows, reports: reports.rows });
    } catch (err) {
        next(err);
    }
});

router.patch('/echoes/:id', authenticate, requireAdmin, async (req, res, next) => {
    try {
        const { status } = req.body;
        if (!['approved', 'hidden'].includes(status)) {
            return res.status(400).json({ error: 'status must be approved or hidden' });
        }
        const result = await query(
            'UPDATE echoes SET status = $1 WHERE id = $2 RETURNING *',
            [status, req.params.id]
        );
        if (!result.rows[0]) return res.status(404).json({ error: 'Echo not found' });
        res.json(result.rows[0]);
    } catch (err) {
        next(err);
    }
});

router.post('/report', authenticate, async (req, res, next) => {
    try {
        const { target_type, target_id, reason } = req.body;
        if (!target_type || !target_id || !reason) {
            return res.status(400).json({ error: 'target_type, target_id, reason are required' });
        }
        if (!['post', 'echo'].includes(target_type)) {
            return res.status(400).json({ error: 'target_type must be post or echo' });
        }
        const result = await query(
            `INSERT INTO reports (reporter_id, target_type, target_id, reason)
             VALUES ($1, $2, $3, $4) RETURNING *`,
            [req.user.id, target_type, target_id, reason]
        );
        res.status(201).json(result.rows[0]);
    } catch (err) {
        next(err);
    }
});

export default router;

File: backend/routes/moderation.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from typing import Optional
import database
from auth_utils import get_current_user, require_admin

router = APIRouter(prefix="/api/moderation", tags=["moderation"])


class ModeratEchoBody(BaseModel):
    status: str


class ReportBody(BaseModel):
    target_type: str
    target_id: str
    reason: str


@router.get("/queue")
async def moderation_queue(admin: dict = Depends(require_admin)):
    posts = await database.query(
        "SELECT id, title, type, status, created_at FROM posts WHERE status = 'pending' ORDER BY created_at ASC"
    )
    echoes = await database.query(
        "SELECT id, post_id, body, status, created_at FROM echoes WHERE status = 'pending' ORDER BY created_at ASC"
    )
    reports = await database.query(
        "SELECT id, target_type, target_id, reason, status, created_at FROM reports WHERE status = 'open' ORDER BY created_at ASC"
    )
    return {
        "posts": [dict(r) for r in posts],
        "echoes": [dict(r) for r in echoes],
        "reports": [dict(r) for r in reports],
    }


@router.patch("/echoes/{echo_id}")
async def moderate_echo(echo_id: str, body: ModeratEchoBody, admin: dict = Depends(require_admin)):
    if body.status not in ("approved", "hidden"):
        raise HTTPException(400, "status must be approved or hidden")
    row = await database.fetchrow(
        "UPDATE echoes SET status = $1 WHERE id = $2 RETURNING *",
        body.status, echo_id,
    )
    if not row:
        raise HTTPException(404, "Echo not found")
    return dict(row)


@router.post("/report", status_code=201)
async def file_report(body: ReportBody, current_user: dict = Depends(get_current_user)):
    if not body.target_type or not body.target_id or not body.reason:
        raise HTTPException(400, "target_type, target_id, reason are required")
    if body.target_type not in ("post", "echo"):
        raise HTTPException(400, "target_type must be post or echo")

    row = await database.fetchrow(
        """INSERT INTO reports (reporter_id, target_type, target_id, reason)
           VALUES ($1, $2, $3, $4) RETURNING *""",
        current_user["id"], body.target_type, body.target_id, body.reason,
    )
    return dict(row)
